pkg/images/enricher/enricher_impl.go

```diff
@@ -208,7 +208,12 @@ func (e *enricherImpl) updateImageWithExistingImage(image *storage.Image, existi
 		return false
 	}
 
-	image.Metadata = existingImage.GetMetadata()
+	// It is possible for us to have a nil metadata stored in case the initial enrichment failed and the image
+	// was referenced by digest. If at a later point a re-scan of the image does contain metadata, we should
+	// retain it (update will happen _after_ fetching metadata).
+	if existingImage.GetMetadata() != nil {
+		image.Metadata = existingImage.GetMetadata()
+	}
```

Comment:
Wondering if this will prevent the metadata from ever being updated after the first successful save. Thinking condition should instead be:

```suggestion
	if image.GetMetadata() == nil {
		image.Metadata = existingImage.GetMetadata()
	}
```

Reply:
I suppose it would since any fetch option disregarding caches would overwrite things, but I've changed it to this one.

Reply:
Actually, second thought... believe the modified logic will now break `delegateEnrichImage`.

`delegateEnrichImage` does not use the metadatacache, it assumes what is in the DB is accurate.  However, `EnrichImage` pulls metadata from the metadatacache, and assumes what is the cache (or freshly pulled) is accurate.  Believe the two use cases require different metadata assignment.

Reply:
Hmm, I'm not sure I get things - in the `delegate` case we want to always update with things from DB, in the `normal enricher` case we want to always use the existing metadata, if not nil, but are also somewhat fine to use the stored metadata (since, effectively, this was how things operated before). 

So `existingImage.GetMetadata() != nil -> overwrite` would be the simplest solution, no? I don't see a world where the metadata cache has new metadata values for the _same_ digest of an image.

Reply:
Your right that the original way you had it is 'better' - although I'm not aware of a scenario in which metadata can be updated - by making this change the only way to update metadata in the future will be to delete the image from the DB, which we may regret if some scenario is discovered.  I'd rather the updates be organic / self healing.  Either way, the condition is an improvement over current state - so am OK moving forward.

Reply:
>by making this change the only way to update metadata in the future will be to delete the image from the DB
I don't think this one is true, during reprocessing we use the fetchoption `IgnoreExistingImage` which will then lead to an update of the image metadata.

Reply:
Looks like reprocessing runs with `ForceRefetchCachedValuesOnly`

https://github.com/stackrox/stackrox/blob/bfa2b597b01a131676ba087600d0e93957001fe9/central/reprocessor/reprocessor.go#L532-L534